datastore/additional_data/sources/nspl.py

```diff
@@ -0,0 +1,166 @@
+
+        return postcodes
+
+    def save_nspl_data(self, data):
+        for record in data:
+            nspl_postcode = (
+                record["pcd"]
+                if record["pcd"]
+                else record["pcd2"]
+                if record["pcd2"]
+                else record["pcds"]
+            )
+            postcode = "".join(nspl_postcode.split()).upper()
+            NSPL.objects.create(postcode=postcode, data=record).save()
+
+    def import_nspl(self, url=NSPL_URL):
+        """
+        Example of a data db entry:
+        {
+            'ccg': 'S03000012', 'ced': None, 'cty': None, 'eer': 'S15000001', 'imd': 6808, 'lat': 57.101474,
+            'pcd': 'AB1 0AA', 'pct': 'S03000012', 'pfa': 'S23000009', 'rgn': None, 'stp': None, 'ctry': 'S92000003',
+            'hash': '3b7db6a231babf449f1c6d7a61ce317f', 'laua': 'S12000033', 'lep1': None, 'lep2': None,
+            'long': -2.242851, 'nuts': 'S31000935', 'oa11': 'S00090303', 'park': None, 'pcd2': 'AB1  0AA',
+            'pcds': 'AB1 0AA', 'pcon': 'S14000002', 'ttwa': 'S22000047', 'ward': 'S13002843', 'wz11': 'S34002990',
+            'bua11': None, 'nhser': None, 'oac11': '1C3', 'calncv': None, 'dointr': '1980-01-01 00:00:00',
+            'doterm': '1996-06-01 00:00:00', 'hlthau': 'S08000020', 'lsoa11': 'S01006514', 'msoa11': 'S02001237',
+            'teclec': 'S09000001', 'buasd11': None, 'ru11ind': '3', 'location': {'lat': 57.101474, 'lon': -2.242851},
+            'oseast1m': 385386, 'osgrdind': 1, 'osnrth1m': 801193, 'usertype': 0
+        }
+        Field names information can be found in NSPL User Guide at https://geoportal.statistics.gov.uk/
+        """
+        zip_file = self.get_zipfile(url)
+        data = self.get_location_data(zip_file)
+        data_json = json.loads(json.dumps(data))
+
+        if NSPL.objects.exists():
+            NSPL.objects.all().delete()
+        self.save_nspl_data(data_json)
+
+    def get_location_data_by_postcode(self, postcode):
+        format_postcode = "".join(postcode.split()).upper()
+        try:
+            nspl = NSPL.objects.get(postcode=format_postcode)
+            return nspl.data
+        except NSPL.DoesNotExist:
+            pass
+
+    def get_location_data_code_names(self, location_data):
+        """Adds new entries to location data dict with code names."""
+        for field_name, field_value in location_data.copy().items():
+            try:
+                code_name_obj = CodeName.objects.get(code=field_value)
+            except CodeName.DoesNotExist:
+                continue
+
+            code_name = code_name_obj.data.get("name")
+            location_data["{}_name".format(field_name)] = code_name
+
+        return location_data
+
+    def update_additional_data(self, grant, additional_data):
+        """Updates with 'recipientOrganizationLocation' based on it's postcode."""
+        recipient_orgs = grant.get("recipientOrganization", [])
```

Comment:
turning this into:

```python
recipient_orgs = grant.get("recipientOrganization", []) + additional_data.get("recipientOrgInfos", [])
```

would allow the postcodes to be used from the regulatory records if a recipient postcode wasn't found.

Reply:
This is now in the code.